packages/drpc/src/DrpcApi.ts

```diff
@@ -0,0 +1,152 @@
+    this.registerMessageHandlers(messageHandlerRegistry)
+  }
+
+  /**
+   * sends the request object to the connection and returns a function that will resolve to the response
+   * @param connectionId the connection to send the request to
+   * @param request the request object
+   * @returns curried function that waits for the response
+   */
+  public async sendRequest(connectionId: string, request: DrpcRequest): Promise<() => Promise<DrpcResponse>> {
+    const connection = await this.connectionService.getById(this.agentContext, connectionId)
+    const { message: drpcMessage, record: drpcMessageRecord } = await this.drpcMessageService.createRequestMessage(
+      this.agentContext,
+      request,
+      connection.id
+    )
+    const messageId = drpcMessage.id
+    await this.sendMessage(connection, drpcMessage, drpcMessageRecord)
+    return this.recvResponse.bind(this, messageId)
+  }
+
+  /**
+   * Listen for a response that has a thread id matching the provided messageId
+   * @param messageId the id to match the response to
+   * @returns the response object
+   */
+  private async recvResponse(messageId: string): Promise<DrpcResponse> {
+    return new Promise((resolve) => {
+      const listener = ({
+        drpcMessageRecord,
+        removeListener,
+      }: {
+        drpcMessageRecord: DrpcMessageRecord
+        removeListener: () => void
+      }) => {
+        const message = drpcMessageRecord.message
+        if (drpcMessageRecord.threadId === messageId) {
+          removeListener()
+          resolve(message as DrpcResponse)
+        }
+      }
+
+      this.drpcMessageService.createResponseListener(listener)
+    })
+  }
+
+  /**
+   * Listen for a request and returns the request object and a function to send the response
+   * @returns the request object and a function to send the response
+   */
+  public async recvRequest(): Promise<{
+    request: DrpcRequest
+    sendResponse: (response: DrpcResponse) => Promise<void>
+  }> {
+    return new Promise((resolve) => {
+      const listener = ({
+        drpcMessageRecord,
+        removeListener,
+      }: {
+        drpcMessageRecord: DrpcMessageRecord
+        removeListener: () => void
+      }) => {
+        const message = drpcMessageRecord.message
+        removeListener()
+        resolve({
+          sendResponse: async (response: DrpcResponse) => {
+            await this.sendResponse(drpcMessageRecord.connectionId, drpcMessageRecord.threadId, response)
+          },
+          request: message as DrpcRequest,
+        })
+      }
+
+      this.drpcMessageService.createRequestListener(listener)
+    })
+  }
+
+  /**
+   * Sends a drpc response to a connection
+   * @param connectionId the connection id to use
+   * @param threadId the thread id to respond to
+   * @param response the drpc response object to send
+   */
+  private async sendResponse(connectionId: string, threadId: string, response: DrpcResponse): Promise<void> {
```

Comment:
It might be a good idea to use an options object here, to avoid unnecessary breaking changes to the API in the future. Something like:

 ```suggestion
  private async sendResponse(options: { connectionId: string, threadId: string, response: DrpcResponse }): Promise<void> {
```